src/lib/storage/segment_accessor.hpp

```diff
@@ -52,37 +54,45 @@ class SegmentAccessor : public BaseSegmentAccessor<T> {
 
 /**
  * For ReferenceSegments, we don't use the SegmentAccessor but either the MultipleChunkReferenceSegmentAccessor or the.
- * SingleChunkReferenceSegmentAccessor. The first one is generally applicable. For each offset that is accessed, a new
- * accessor has to be created. This is because we cannot be sure that two consecutive offsets reference the same chunk.
- * In the SingleChunkReferenceSegmentAccessor, we know that the same chunk is referenced, so we create the accessor
- * only once.
+ * SingleChunkReferenceSegmentAccessor. The first one is generally applicable. However, we will have more overhead,
+ * because we cannot be sure that two consecutive offsets reference the same chunk. In the
+ * SingleChunkReferenceSegmentAccessor, we know that the same chunk is referenced, so we create the accessor only once.
  */
 template <typename T>
-class MultipleChunkReferenceSegmentAccessor : public BaseSegmentAccessor<T> {
+class MultipleChunkReferenceSegmentAccessor : public AbstractSegmentAccessor<T> {
  public:
-  explicit MultipleChunkReferenceSegmentAccessor(const ReferenceSegment& segment) : _segment{segment} {}
+  explicit MultipleChunkReferenceSegmentAccessor(const ReferenceSegment& segment)
+      : _segment{segment}, _table{segment.referenced_table()}, _accessors{1} {}
 
   const std::optional<T> access(ChunkOffset offset) const final {
-    const auto& referenced_row_id = (*_segment.pos_list())[offset];
-    if (referenced_row_id.is_null()) return std::nullopt;
+    const auto& row_id = (*_segment.pos_list())[offset];
+    if (row_id.is_null()) {
+      return std::nullopt;
+    }
 
-    const auto& table = _segment.referenced_table();
-    const auto referenced_column_id = _segment.referenced_column_id();
-    const auto referenced_chunk_id = referenced_row_id.chunk_id;
-    const auto referenced_chunk_offset = referenced_row_id.chunk_offset;
+    const auto chunk_id = row_id.chunk_id;
 
-    const auto accessor =
-        create_segment_accessor<T>(table->get_chunk(referenced_chunk_id)->get_segment(referenced_column_id));
-    return accessor->access(referenced_chunk_offset);
+    if (static_cast<size_t>(chunk_id) >= _accessors.size()) {
```

Comment:
Is this check necessary? Do we not always know the number of chunks in advance?

Reply:
As the number of potential Chunks is not part of the create_segment_accessor interface, this check is necessary. I don't want to change the interface of the function for this.

Reply:
Am on a mobile, so I cannnot check the code, but don’t we pass the table and can thus directly get the chunk count? If not, please ignore.

Reply:
Yeah, COULD be done, but the whole array of one-accessor-for-each-chunk thing is such a (unavoidable, atm :( ) waste that I doubt it would matter much if we safe a few resizes and instead allocate a longer vector than we might actually need. I'd leave it as it is.